server/src/main/java/org/candlepin/dto/api/v1/EnvironmentDTO.java

```diff
@@ -0,0 +1,276 @@
+     *
+     * @return a reference to this DTO object.
+     */
+    public EnvironmentDTO setEnvironmentContent(Set<EnvironmentContentDTO> environmentContent) {
+        if (environmentContent != null) {
+            if (this.environmentContent == null) {
+                this.environmentContent = new HashSet<EnvironmentContentDTO>();
+            }
+            else {
+                this.environmentContent.clear();
+            }
+
+            for (EnvironmentContentDTO dto : environmentContent) {
+                if (dto == null || dto.getContent() == null ||
+                    StringUtils.isEmpty(dto.getContent().getId())) {
+                    throw new IllegalArgumentException("environment content is null or incomplete");
+                }
+
+                this.environmentContent.add(new EnvironmentContentDTO(dto));
+            }
+        }
+        else {
+            this.environmentContent = null;
+        }
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return String.format(
+            "EnvironmentDTO [id: %s, name: %s]",
+            this.getId(), this.getName());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        }
+
+        if (obj instanceof EnvironmentDTO && super.equals(obj)) {
+            EnvironmentDTO that = (EnvironmentDTO) obj;
+
+            String thisOid = this.getOwner() != null ? this.getOwner().getId() : null;
+            String thatOid = that.getOwner() != null ? that.getOwner().getId() : null;
+
+            EqualsBuilder builder = new EqualsBuilder()
+                .append(this.getId(), that.getId())
+                .append(this.getName(), that.getName())
+                .append(this.getDescription(), that.getDescription())
+                .append(thisOid, thatOid);
+
+            return builder.isEquals() &&
+                Util.collectionsAreEqual(this.getEnvironmentContent(), that.getEnvironmentContent());
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        String thisOid = this.getOwner() != null ? this.getOwner().getId() : null;
+
+        HashCodeBuilder builder = new HashCodeBuilder(37, 7)
+            .append(super.hashCode())
+            .append(this.getId())
+            .append(this.getName())
+            .append(this.getDescription())
+            .append(thisOid)
+            .append(this.getEnvironmentContent());
+
+        return builder.toHashCode();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public EnvironmentDTO clone() {
+        EnvironmentDTO copy = super.clone();
+        copy.owner = owner != null ? owner.clone() : null;
+        copy.setEnvironmentContent(this.getEnvironmentContent());
```

Comment:
This is only acceptable if, and only if, the EnvironmentContentDTO is immutable (which it does not appear to be). This is okay if the implementation there is changed to closer match ProductContentDTO.

Reply:
changed to mimic ProductContentDTO